crates/ruff_linter/src/rules/flake8_pyi/rules/unnecessary_literal_union.rs

```diff
@@ -32,43 +34,163 @@ pub struct UnnecessaryLiteralUnion {
 }
 
 impl Violation for UnnecessaryLiteralUnion {
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;
+
     #[derive_message_formats]
     fn message(&self) -> String {
         format!(
             "Multiple literal members in a union. Use a single literal, e.g. `Literal[{}]`",
             self.members.join(", ")
         )
     }
+
+    fn fix_title(&self) -> Option<String> {
+        Some(format!("Replace with a single `Literal`",))
+    }
+}
+
+fn concatenate_bin_ors(exprs: Vec<&Expr>) -> Expr {
+    let mut exprs = exprs.into_iter();
+    let first = exprs.next().unwrap();
+    exprs.fold((*first).clone(), |acc, expr| {
+        Expr::BinOp(ast::ExprBinOp {
+            left: Box::new(acc),
+            op: Operator::BitOr,
+            right: Box::new((*expr).clone()),
+            range: TextRange::default(),
+        })
+    })
+}
+
+fn make_union(subscript: &ExprSubscript, exprs: Vec<&Expr>) -> Expr {
+    Expr::Subscript(ast::ExprSubscript {
+        value: subscript.value.clone(),
+        slice: Box::new(Expr::Tuple(ast::ExprTuple {
+            elts: exprs.into_iter().map(|expr| (*expr).clone()).collect(),
+            range: TextRange::default(),
+            ctx: ast::ExprContext::Load,
+        })),
+        range: TextRange::default(),
+        ctx: ast::ExprContext::Load,
+    })
+}
+
+fn make_literal_expr(subscript: Option<Expr>, exprs: Vec<&Expr>) -> Expr {
+    let use_subscript = if let subscript @ Some(_) = subscript {
+        subscript.unwrap().clone()
+    } else {
+        Expr::Name(ast::ExprName {
+            id: "Literal".to_string(),
+            range: TextRange::default(),
+            ctx: ast::ExprContext::Load,
+        })
+    };
+    Expr::Subscript(ast::ExprSubscript {
+        value: Box::new(use_subscript),
+        slice: Box::new(Expr::Tuple(ast::ExprTuple {
+            elts: exprs.into_iter().map(|expr| (*expr).clone()).collect(),
+            range: TextRange::default(),
+            ctx: ast::ExprContext::Load,
+        })),
+        range: TextRange::default(),
+        ctx: ast::ExprContext::Load,
+    })
 }
 
 /// PYI030
 pub(crate) fn unnecessary_literal_union<'a>(checker: &mut Checker, expr: &'a Expr) {
     let mut literal_exprs = Vec::new();
+    let mut other_exprs = Vec::new();
 
-    // Adds a member to `literal_exprs` if it is a `Literal` annotation
+    // for the sake of consistency and correctness, we'll use the first Literal subscript attribute
+    let mut literal_subscript = None;
+    let mut total_literals = 0;
+
+    // Adds a member to `literal_exprs` if it is a `Literal` annotation.
```

Comment:
Should probably include note about `other_exprs` to e.g.
```suggestion
    // Split members into `literal_exprs` if they are a `Literal` annotation  and `other_exprs` otherwise
```